src/test/java/com/commercetools/sync/customobjects/CustomObjectSyncTest.java

```diff
@@ -0,0 +1,407 @@
+        final List<String> errorMessages = new ArrayList<>();
+        final List<Throwable> exceptions = new ArrayList<>();
+
+        final CustomObjectSyncOptions spyCustomObjectSyncOptions =
+                initCustomObjectSyncOptions(errorMessages, exceptions);
+
+        final CustomObject<JsonNode> existingCustomObject = mock(CustomObject.class);
+        when(existingCustomObject.getContainer()).thenReturn("someContainer");
+        when(existingCustomObject.getKey()).thenReturn("someKey");
+        when(existingCustomObject.getValue()).thenReturn(JsonNodeFactory.instance.numberNode(2020));
+
+        final Set<CustomObject<JsonNode>> existingCustomObjectSet = new HashSet<CustomObject<JsonNode>>();
+        existingCustomObjectSet.add(existingCustomObject);
+
+        final CustomObject<JsonNode> updatedCustomObject = mock(CustomObject.class);
+        when(updatedCustomObject.getContainer()).thenReturn("someContainer");
+        when(updatedCustomObject.getKey()).thenReturn("someKey");
+        when(updatedCustomObject.getValue()).thenReturn(newCustomObjectDraft.getValue());
+
+        final CustomObjectService customObjectService = mock(CustomObjectService.class);
+        when(customObjectService.fetchMatchingCustomObjects(anySet()))
+                .thenReturn(completedFuture(existingCustomObjectSet));
+        when(customObjectService.upsertCustomObject(any()))
+                .thenReturn(supplyAsync(() -> { throw new SphereException(); }));
+        when(customObjectService.fetchCustomObject(any(CustomObjectCompositeIdentifier.class)))
+                .thenReturn(supplyAsync(() -> {
+                    throw new SphereException();
+                }));
+
+        // test
+        CustomObjectSyncStatistics syncStatistics =
+                new CustomObjectSync(spyCustomObjectSyncOptions, customObjectService)
+                        .sync(singletonList(newCustomObjectDraft)).toCompletableFuture().join();
+
+        // assertion
+        assertAll(
+            () -> assertThat(syncStatistics.getProcessed().get()).isEqualTo(1),
+            () -> assertThat(syncStatistics.getCreated().get()).isEqualTo(0),
+            () -> assertThat(syncStatistics.getUpdated().get()).isEqualTo(0),
+            () -> assertThat(syncStatistics.getFailed().get()).isEqualTo(1)
+        );
+        assertThat(exceptions).hasSize(1);
+        assertThat(errorMessages)
+            .hasSize(1)
+            .singleElement()
+            .isEqualTo(
+                    format("Failed to update custom object with key: '%s'. Reason: %s",
+                            CustomObjectCompositeIdentifier.of(newCustomObjectDraft).toString(),
+                            exceptions.get(0).getMessage())
+            );
+        verify(customObjectService).upsertCustomObject(any());
+        verify(customObjectService).fetchMatchingCustomObjects(any());
+    }
+
+    @Test
+    void sync_WithDifferentIdentifiers_ShouldCreateSuccessfully() {
```

Comment:
can you add a test case for same container but different keys scenario? and also same key but different containers

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/569/commits/5626de036e06f9b1c9464abeed4c7651f66aa382